client/mysqlbinlog.cc

```diff
@@ -2922,6 +2923,17 @@ static Exit_status dump_local_log_entries(PRINT_EVENT_INFO *print_event_info,
 
     Log_event* ev = Log_event::read_log_event(file, glob_description_event,
                                               opt_verify_binlog_checksum);
+
+    // If we're not in raw mode, then emit a warning in the event that we
+    // finished processing input before reaching the boundary indicated by
+    // --stop-position.
+    if (!opt_raw_mode &&  // --stop-position is ignored in raw mode
```

Comment:
We shouldn't need to check for each event seen, but we should only check once after we have processed everything.  `process_event()` will return `OK_STOP` if it has encountered the stop position, and this `Exit_status` value should propagate up to main. So after we've dumped events from all files, we should be able to check if `stop_position` (and for consistency, probably `stop_datetime` too, as a last-minute thought :sweat_smile: And can you add a test for that too?) is set and `retval != OK_STOP`, then we warn.

Reply:
I see: we will know (1) if `stop_position` was set by the user and (2) whether we hit EOF first or `stop_position` first as in the latter case we will see `OK_STOP`; then we can deduce if we need to warn or not at the top-level.  The only trouble with this approach is that we cannot distinguish between when we've read exactly to EOF or when we've read past it, both cases will appear the same.  This is because `OK_STOP` is returned when the position is greater-than or equal to `stop_position`.  We could disambiguate with for example `OK_STOP_BEFORE_EOF` and `OK_STOP_AT_EOF`.
Yes, I'll add checks and tests for `stop_datetime` as well 👍

Reply:
Great observation! I don't necessarily think we need the new status options though. I think we can just override `retval` to `OK_STOP` in such case. E.g., in `dump_local_log_entries`, at the line

```
      // file->error == 0 means EOF, that's OK, we break in this case
```
we check if we have hit our stop condition, and if so, set `retval= OK_STOP`. It seems more consistent this way too because stopping should be the same whether we are at or before the EOF (note `OK_EOF` isn't actually used for `dump_local_log_entries`; the alternative is `OK_CONTINUE`)

Reply:
If we override `retval` to `OK_STOP` and in the case that `end_io_cache` returns a failure, we will inadvertently swallow the warning because `retval` will become `ERROR_STOP` (ed of `dump_local_log_entries`).  That may be fine for the user, but I'm not sure what the user expectations are; in the case of an error, should the user see any/all warnings that may have occurred?

Reply:
Yeah it'd be better not to swallow the warning, good point. And with the prior amendment where we are already adding a check after processing each binlog at
```
      // file->error == 0 means EOF, that's OK, we break in this case
```
we may as well write the warning there and avoid a redundant later check in `main`. We could still set `retval= OK_STOP` too, it seems more consistent.

Reply:
I was just looking at doing that, we're on the same brainwave 😄